src/mlpack/methods/ann/layer/weight_norm.hpp

```diff
@@ -0,0 +1,204 @@
+   * It then calulates the output of the wrapped layer from the calculated
+   * weights.
+   *
+   * @param input Input data for the layer.
+   * @param output Resulting output activations.
+   */
+  template<typename eT>
+  void Forward(arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Backward pass through the layer. This function calls the Backward()
+   * function of the wrapped layer.
+   *
+   * @param input The input activations.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& input,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /**
+   * Calculate the gradient using the output delta and the input activations and
+   * weights of the wrapped layer.
+   *
+   * @param input The input activations.
+   * @param error The calculated error.
+   * @param gradient The calculated gradient.
+   */
+  template<typename eT>
+  void Gradient(arma::Mat<eT>&& input,
+                arma::Mat<eT>&& error,
+                arma::Mat<eT>&& gradient);
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+  //! Modify the wrapped layer.
+  LayerTypes<CustomLayers...>& Layer() { return wrappedLayer; }
+
+  /**
+   * Serialize the layer.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored number of bias elements in the weights of wrapped layer.
+  size_t biasWeightSize;
+
+  //! Locally-stored delete visitor module object.
+  DeleteVisitor deleteVisitor;
+
+  //! Locally-stored delta object.
+  OutputDataType delta;
+
+  //! Locally-stored delta visitor module object.
+  DeltaVisitor deltaVisitor;
+
+  //! Locally-stored gradient object.
+  OutputDataType gradient;
+
+  //! Locally-stored wrapped layer.
+  LayerTypes<CustomLayers...> wrappedLayer;
+
+  //! Locally stored number of elements in the weights of wrapped layer.
+  size_t layerWeightSize;
+
+  //! Locally-stored output parameter object.
+  OutputDataType outputParameter;
+
+  //! Locally-stored output parameter visitor module object.
+  OutputParameterVisitor outputParameterVisitor;
+
+  //! Reset the gradient for all modules that implement the Gradient function.
+  void ResetGradients(arma::mat& gradient);
```

Comment:
Why is this function made private?

Reply:
I have made this private taking `line 368` in `ffn.hpp` as a reference. It is also made private so that it is `not accessible directly`.